flink/src/test/java/org/apache/iceberg/flink/TestFlinkTableSource.java

```diff
@@ -103,4 +124,45 @@ public void testLimitPushDown() {
     Assert.assertEquals("should have 1 record", 1, mixedResult.size());
     Assert.assertArrayEquals("Should produce the expected records", mixedResult.get(0), new Object[] {1, "a"});
   }
+
+  @Test
+  public void testParallelismOptimize() {
+    sql("INSERT INTO %s  VALUES (1,'hello')", TABLE_NAME);
+    sql("INSERT INTO %s  VALUES (2,'iceberg')", TABLE_NAME);
+
+    TableEnvironment tenv = getTableEnv();
+
+    // empty table ,parallelism at least 1
+    Table tableEmpty = tenv.sqlQuery(String.format("SELECT * FROM %s", TABLE_NAME));
+    testParallelismSettingTranslateAndAssert(1, tableEmpty, tenv);
+
+    // make sure to generate 2 CombinedScanTasks
+    org.apache.iceberg.Table table = validationCatalog.loadTable(TableIdentifier.of(icebergNamespace, TABLE_NAME));
+    Stream<FileScanTask> stream = StreamSupport.stream(table.newScan().planFiles().spliterator(), false);
+    Optional<FileScanTask> fileScanTaskOptional =  stream.max(Comparator.comparing(FileScanTask::length));
+    Assert.assertTrue(fileScanTaskOptional.isPresent());
+    long maxFileLen = fileScanTaskOptional.get().length();
+    sql("ALTER TABLE %s SET ('read.split.open-file-cost'='1', 'read.split.target-size'='%s')", TABLE_NAME, maxFileLen);
+
+    // 2 splits ,the parallelism is  2
+    Table tableSelect = tenv.sqlQuery(String.format("SELECT * FROM %s", TABLE_NAME));
+    testParallelismSettingTranslateAndAssert(2, tableSelect, tenv);
+
+    // 2 splits  and limit is 1 ,the parallelism is  1
+    Table tableLimit = tenv.sqlQuery(String.format("SELECT * FROM %s LIMIT 1", TABLE_NAME));
+    testParallelismSettingTranslateAndAssert(1, tableLimit, tenv);
+  }
+
+  private void testParallelismSettingTranslateAndAssert(int expected, Table table, TableEnvironment tEnv) {
```

Comment:
Is there another way to assert the parallelism as expected value ?  Here we're using flink's planner to get the `ExecNode` ,  I'm concerning that we're using flink's `Internal` codes which would be a big trouble when upgrading the flink version.  Pls see this PR https://github.com/apache/iceberg/pull/1956

Reply:
Here we need to get the parallelism of an Operator. It is indeed get by using the flink internal code, because I have not found other ways to obtain the parallelism.

Reply:
Flink 1.12 does refactor `ExecNode`, I found an easier way to assert parallelism, I will update it later